### PR TITLE
implementation of custom fields for terraform

### PR DESCRIPTION
**Pull request by jedelson-pagerduty**
This adds support for the new PagerDuty Custom Fields on Incidents features. These are currently in Early Access and not available to all accounts.

API Documentation for this feature are visible on https://developer.pagerduty.com/api-reference/f1a95bb9397ba-changelog

Dependent upon https://github.com/heimweh/go-pagerduty/pull/120

Testing is done similarly to incident workflows where there is an environment variable (`PAGERDUTY_ACC_CUSTOM_FIELDS`) that must be set. As with workflows, any value will do, but `1` is recommended:

```
PAGERDUTY_ACC_CUSTOM_FIELDS=1 make testacc TESTARGS="-run PagerDutyCustomField"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run PagerDutyCustomField -timeout 120m
?       github.com/terraform-providers/terraform-provider-pagerduty     [no test files]
=== RUN   TestAccDataSourcePagerDutyCustomField
--- PASS: TestAccDataSourcePagerDutyCustomField (6.24s)
=== RUN   TestAccDataSourcePagerDutyCustomFieldSchema
--- PASS: TestAccDataSourcePagerDutyCustomFieldSchema (5.20s)
=== RUN   TestAccPagerDutyCustomFieldOptions_Basic
--- PASS: TestAccPagerDutyCustomFieldOptions_Basic (11.44s)
=== RUN   TestAccPagerDutyCustomFieldOptions_Integer
--- PASS: TestAccPagerDutyCustomFieldOptions_Integer (11.48s)
=== RUN   TestAccPagerDutyCustomFieldOptions_Integer_Bad
--- PASS: TestAccPagerDutyCustomFieldOptions_Integer_Bad (1.60s)
=== RUN   TestAccPagerDutyCustomFieldOptions_Integer_Bad_Float
--- PASS: TestAccPagerDutyCustomFieldOptions_Integer_Bad_Float (1.22s)
=== RUN   TestAccPagerDutyCustomFieldOptions_Float
--- PASS: TestAccPagerDutyCustomFieldOptions_Float (11.48s)
=== RUN   TestAccPagerDutyCustomFieldOptions_Float_Bad
--- PASS: TestAccPagerDutyCustomFieldOptions_Float_Bad (1.10s)
=== RUN   TestAccPagerDutyCustomFieldSchemaAssignment
--- PASS: TestAccPagerDutyCustomFieldSchemaAssignment (12.06s)
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_default!!!_for_type_integer,_multi-value:_false
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_42_for_type_integer,_multi-value:_false
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_42_for_type_integer,_multi-value:_true
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_[42,"foo"]_for_type_integer,_multi-value:_true
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_[50.5,"foo"]_for_type_float,_multi-value:_true
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_[50.5,42.3]_for_type_float,_multi-value:_true
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_default!!!_for_type_float,_multi-value:_false
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_50.5_for_type_float,_multi-value:_false
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_50.5_for_type_float,_multi-value:_true
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_[50.5,"foo"]_for_type_float,_multi-value:_true#01
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_[50.5,42.3]_for_type_float,_multi-value:_true#01
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_default!!!_for_type_boolean,_multi-value:_false
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_True_for_type_boolean,_multi-value:_false
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_true_for_type_boolean,_multi-value:_false
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_false_for_type_boolean,_multi-value:_false
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_default!!!_for_type_url,_multi-value:_false
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_\x01_for_type_url,_multi-value:_false
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_https://www.pagerduty.com/_for_type_url,_multi-value:_false
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_default!!!_for_type_datetime,_multi-value:_false
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_2022-01-03T15:04:05Z_for_type_datetime,_multi-value:_false
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_anything_goes_--_validated_server-side_for_type_field_option,_multi-value:_false
=== RUN   TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_something_for_type_field_option,_multi-value:_true
--- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_default!!!_for_type_integer,_multi-value:_false (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_42_for_type_integer,_multi-value:_false (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_42_for_type_integer,_multi-value:_true (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_[42,"foo"]_for_type_integer,_multi-value:_true (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_[50.5,"foo"]_for_type_float,_multi-value:_true (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_[50.5,42.3]_for_type_float,_multi-value:_true (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_default!!!_for_type_float,_multi-value:_false (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_50.5_for_type_float,_multi-value:_false (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_50.5_for_type_float,_multi-value:_true (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_[50.5,"foo"]_for_type_float,_multi-value:_true#01 (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_[50.5,42.3]_for_type_float,_multi-value:_true#01 (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_default!!!_for_type_boolean,_multi-value:_false (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_True_for_type_boolean,_multi-value:_false (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_true_for_type_boolean,_multi-value:_false (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_false_for_type_boolean,_multi-value:_false (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_default!!!_for_type_url,_multi-value:_false (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_\x01_for_type_url,_multi-value:_false (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_https://www.pagerduty.com/_for_type_url,_multi-value:_false (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_default!!!_for_type_datetime,_multi-value:_false (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_2022-01-03T15:04:05Z_for_type_datetime,_multi-value:_false (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_anything_goes_--_validated_server-side_for_type_field_option,_multi-value:_false (0.00s)
    --- PASS: TestPagerDutyCustomField_ValidateDefaultFieldValue/Test_value:_something_for_type_field_option,_multi-value:_true (0.00s)
=== RUN   TestAccPagerDutyCustomFieldConfiguration_Basic
--- PASS: TestAccPagerDutyCustomFieldConfiguration_Basic (7.31s)
=== RUN   TestAccPagerDutyCustomFieldConfiguration_Required
--- PASS: TestAccPagerDutyCustomFieldConfiguration_Required (12.79s)
=== RUN   TestAccPagerDutyCustomFieldConfiguration_Required_BadDataType
--- PASS: TestAccPagerDutyCustomFieldConfiguration_Required_BadDataType (1.04s)
=== RUN   TestAccPagerDutyCustomFieldConfiguration_Required_Without_Default
--- PASS: TestAccPagerDutyCustomFieldConfiguration_Required_Without_Default (1.02s)
=== RUN   TestAccPagerDutyCustomFieldConfiguration_Required_Without_DefaultDataType
--- PASS: TestAccPagerDutyCustomFieldConfiguration_Required_Without_DefaultDataType (1.00s)
=== RUN   TestAccPagerDutyCustomFieldConfiguration_Default_Without_Required
--- PASS: TestAccPagerDutyCustomFieldConfiguration_Default_Without_Required (1.03s)
=== RUN   TestAccPagerDutyCustomFieldConfiguration_Required_InvalidValue_MultiValue
--- PASS: TestAccPagerDutyCustomFieldConfiguration_Required_InvalidValue_MultiValue (1.08s)
=== RUN   TestAccPagerDutyCustomFieldSchemas_Basic
--- PASS: TestAccPagerDutyCustomFieldSchemas_Basic (8.16s)
=== RUN   TestAccPagerDutyCustomFields_Basic
--- PASS: TestAccPagerDutyCustomFields_Basic (8.09s)
=== RUN   TestAccPagerDutyCustomField_BasicWithDescription
--- PASS: TestAccPagerDutyCustomField_BasicWithDescription (4.42s)
=== RUN   TestAccPagerDutyCustomFields_UnknownDataType
--- PASS: TestAccPagerDutyCustomFields_UnknownDataType (1.03s)
=== RUN   TestAccPagerDutyCustomFields_IllegalDataType
--- PASS: TestAccPagerDutyCustomFields_IllegalDataType (0.97s)
=== RUN   TestPagerDutyCustomField_ConvertValueForBuild
--- PASS: TestPagerDutyCustomField_ConvertValueForBuild (0.00s)
=== RUN   TestPagerDutyCustomField_ConvertDefaultValueForFlatten
--- PASS: TestPagerDutyCustomField_ConvertDefaultValueForFlatten (0.00s)
PASS
ok      github.com/terraform-providers/terraform-provider-pagerduty/pagerduty   110.134s
```

### CODE REVIEW

File: pagerduty/resource_pagerduty_custom_field.go [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+		FixedOptions: d.Get("fixed_options").(bool),
+	}
+	if desc, ok := d.GetOk("description"); ok {
+		str := desc.(string)
+		field.Description = &str
+	}
+	return &field, nil
+}
+
+func validateFieldDataType() schema.SchemaValidateDiagFunc {
```
Comment (imjaroiswebdev) [inline pagerduty/resource_pagerduty_custom_field.go:189]: Could you please move the utility functions that are being re-used for this resource, on custom field options and schemas to a new file with that purpose? in a fashion like `event_orchestration_path_util.go`

**Review by imjaroiswebdev — CHANGES_REQUESTED** [review]
@jedelson-pagerduty before proceeding with the merge, there are a few comments that need to be addressed please

File: website/docs/r/custom_field.html.markdown [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+layout: "pagerduty"
+page_title: "PagerDuty: pagerduty_custom_field"
+sidebar_current: "docs-pagerduty-resource-custom-field"
+description: |-
+  Creates and manages a custom field in PagerDuty.
+---
+
+# pagerduty\_custom\_field
+
+A Custom Field is a resuable element which can be added to Custom Field Schemas.
```
[the remark this answers is not in the document]
Comment (jedelson-pagerduty) [reply website/docs/r/custom_field.html.markdown]: This is addressed now.

File: website/docs/r/custom_field_schema.html.markdown [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+layout: "pagerduty"
+page_title: "PagerDuty: pagerduty_custom_field_schema"
+sidebar_current: "docs-pagerduty-resource-custom-field-schema"
+description: |-
+  Creates and manages a custom field schema in PagerDuty.
+---
+
+# pagerduty\_custom\_field\_schema
+
+A Custom Field Schema is a set of custom fields which can be set on an incident.
```
[the remark this answers is not in the document]
Comment (jedelson-pagerduty) [reply website/docs/r/custom_field_schema.html.markdown]: I chose a slightly different deep link, but same idea.

File: website/docs/r/custom_field_schema_field_configuration.html.markdown [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+layout: "pagerduty"
+page_title: "PagerDuty: pagerduty_custom_field_schema_field_configuration"
+sidebar_current: "docs-pagerduty-resource-custom-field-schema-field-configuration"
+description: |-
+  Creates and manages a custom field configuration in PagerDuty.
+---
+
+# pagerduty\_custom\_field\_schema\_field\_configuration
+
+A Custom Field Configuration is a declaration of a specific Custom Field in a specific Custom Field Schema.
```
[the remark this answers is not in the document]
Comment (jedelson-pagerduty) [reply website/docs/r/custom_field_schema_field_configuration.html.markdown]: Addressed

File: website/docs/r/custom_field_schema_assignment.html.markdown [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+layout: "pagerduty"
+page_title: "PagerDuty: pagerduty_custom_field_schema_assignment"
+sidebar_current: "docs-pagerduty-resource-custom-field-schema-assignment"
+description: |-
+  Creates and manages a custom field schema assignment in PagerDuty.
+---
+
+# pagerduty\_custom\_field\_schema\_assignment
+
+A Custom Field Schema Assignment is a relationship between a Custom Field Schema and a Service.
```
[the remark this answers is not in the document]
Comment (jedelson-pagerduty) [reply website/docs/r/custom_field_schema_assignment.html.markdown]: yep

File: website/docs/d/custom_field.html.markdown [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    field = data.pagerduty_custom_field.sre_environment.id
+  }
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `namespace` - (Required) The namespace of the field.
```
[the remark this answers is not in the document]
Comment (jedelson-pagerduty) [reply website/docs/d/custom_field.html.markdown]: whoops. thanks for catching this!

File: website/docs/d/custom_field.html.markdown [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+layout: "pagerduty"
+page_title: "PagerDuty: pagerduty_custom_field"
+sidebar_current: "docs-pagerduty-datasource-custom-field"
+description: |-
+  Get information about a custom field that you can add to a custom field schema.
+---
+
+# pagerduty\_custom\_field
+
+Use this data source to get information about a specific custom field that you can add to a custom field schema.
```
[the remark this answers is not in the document]
Comment (jedelson-pagerduty) [reply website/docs/d/custom_field.html.markdown]: yep

File: pagerduty/util.go [hunk 8]
```diff
@@ -200,3 +200,10 @@ func resourcePagerDutyParseColonCompoundID(id string) (string, string) {
 	parts := strings.Split(id, ":")
 	return parts[0], parts[1]
 }
+
+func hashOfMapValue(k string) schema.SchemaSetFunc {
```
[the remark this answers is not in the document]
Comment (jedelson-pagerduty) [reply pagerduty/util.go]: thanks for catching this!

File: pagerduty/resource_pagerduty_custom_field.go (hunk 1 above)
Comment (jedelson-pagerduty) [reply pagerduty/resource_pagerduty_custom_field.go]: Great idea! Hadn't seen this pattern applied in this provider code so didn't want to invent anything new. Moved the functions used in more than one resource to `custom_field_util.go` and the unit tests for those (although most of the utility code is tested via acceptance tests) in `custom_field_util_test.go`.

File: website/docs/r/custom_field_option.html.markdown [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+  datatype = "string"
+  value    = "prod"
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `datatype` - (Required) The datatype of the field. Must be one of `string`, `integer`, `float`, `boolean`, `datetime`, or `url`.
```
[the remark this answers is not in the document]
Comment (jedelson-pagerduty) [reply website/docs/r/custom_field_option.html.markdown]: good catch. thanks!

**Comment by jedelson-pagerduty** [comment]
Thanks for all the feedback @imjaroiswebdev . I think I've addressed all of these issues.

File: pagerduty/resource_pagerduty_custom_field_schema_field_configuration.go [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+		if err := flattenFieldSchemaFieldConfiguration(d, schemaID, fieldConfiguration); err != nil {
+			return resource.NonRetryableError(err)
+		}
+		return nil
+
+	})
+}
+
+func flattenFieldSchemaFieldConfiguration(d *schema.ResourceData, schemaID string, fc *pagerduty.CustomFieldSchemaFieldConfiguration) error {
+	d.SetId(fc.ID)
```
Comment (imjaroiswebdev) [inline pagerduty/resource_pagerduty_custom_field_schema_field_configuration.go:156]: Could you move this `SetId` after the error handling please?

File: pagerduty/resource_pagerduty_custom_field_option.go [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+			},
+			"value": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+		},
+	}
+}
+
+func validateCustomFieldOptionValue(_ context.Context, diff *schema.ResourceDiff, _ interface{}) error {
```
Comment (imjaroiswebdev) [inline pagerduty/resource_pagerduty_custom_field_option.go:40]: Could you please comment on the code what is the scenario needing this customizing diff logic?

**Review by imjaroiswebdev — CHANGES_REQUESTED** [review]
There are still a couple of comments to be addressed before merging @jedelson-pagerduty please 🙏🏽

File: pagerduty/resource_pagerduty_custom_field_schema_field_configuration.go (hunk 10 above)
Comment (jedelson-pagerduty) [reply pagerduty/resource_pagerduty_custom_field_schema_field_configuration.go]: That makes sense. Hadn't thought of that.

File: pagerduty/resource_pagerduty_custom_field_option.go (hunk 11 above)
Comment (jedelson-pagerduty) [reply pagerduty/resource_pagerduty_custom_field_option.go:42]: I put the comment in the schema.Resource definition at the top of this and in `resource_custom_field_schema_field_configuration` (which does something similar)

**Review by imjaroiswebdev — APPROVED** [review]
Awesome! Terraform managed Custom Fields will be loved by users. Thank you @jedelson-pagerduty 🎉 💪🏽 😎

**Pull request merged**